### PR TITLE
Add expiry time to preauthed urls

### LINKED ISSUE #3238: preauthenticated urls should expire
The `ContentRedirectingContentGuard` is able to emit preauthenticated urls. The should have an expiry.

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #3238

### CODE REVIEW

**Review by dkliban — COMMENTED** [review]
What do you think about changing the name of the query parameter to `expiry` instead of `expire`?

**Comment by mdellweg** [comment]
One question that came to me: Do we want the timeout to be configurable? As a setting? Or as a model field?

**Comment by dkliban** [comment]
The other option is calling it `expires`.

**Comment by mdellweg** [comment]
> It would be useful to make the expiration configurable. Who is expected to set this value? The REST API user or the admin? The plugin writer?

Now that is really the question. I can leave it like that for now and then change to either later. But once we have it as a setting, the ship really will have sailed. And with the REST API parameter it's the same the other way around.

File: pulpcore/app/models/publication.py [hunk 1]
```diff
@@ -385,19 +386,24 @@ class ContentRedirectContentGuard(ContentGuard, AutoAddObjPermsMixin):
 
     TYPE = "content_redirect"
 
+    EXPIRETIME = timedelta(minutes=5)
```
Comment (ipanova) [inline pulpcore/app/models/publication.py:389]: what was your judgement around the expiration time? can we set it to 1h? i looked into aws s3 presigned url and they expire by default in 1 hour.

Comment (mdellweg) [reply pulpcore/app/models/publication.py:389]: It's about `pi * thumb`. Sure i can change that to 1h.

File: pulpcore/app/models/publication.py [hunk 2]
```diff
@@ -385,19 +386,24 @@ class ContentRedirectContentGuard(ContentGuard, AutoAddObjPermsMixin):
...
+
     shared_secret = models.BinaryField(max_length=32, default=_gen_secret)
 
     def permit(self, request):
         """
         Permit preauthenticated redirects from pulp-api.
         """
         try:
+            expires = request.query["expires"]
+            if int(datetime.now().timestamp()) > int(expires):
```
Comment (ipanova) [inline pulpcore/app/models/publication.py:399]: should we use instead `from django.utils.timezone import now`

File: pulpcore/app/models/publication.py [hunk 3]
```diff
@@ -410,9 +416,14 @@ def preauthenticate_url(self, url, salt=None):
...
-        digest = self._get_digest(salt, url).hex()
-        url = url + f"?validate_token={hex_salt}:{digest}"
-        return url
+        expiretime = int((datetime.now() + self.EXPIRETIME).timestamp())
+        if "?" in url:
+            separator = "&"
+        else:
+            separator = "?"
+        url_to_sign = url + separator + f"expires={expiretime}"
+        digest = self._get_digest(salt, url_to_sign).hex()
```
Comment (ipanova) [inline pulpcore/app/models/publication.py:425]: I am not sure we need to include the expiration data in the url to sign here.

Comment (mdellweg) [reply pulpcore/app/models/publication.py:425]: We need that absolutely. It must not be tempered with. (There is even a test for it.)

File: pulpcore/app/models/publication.py (hunk 2 above)
Comment (mdellweg) [reply pulpcore/app/models/publication.py:399]: https://docs.djangoproject.com/en/3.2/topics/i18n/timezones/
This is hard to digest. But probably yes. So you cannot bypass the check by specifying a weird timezome.

**Review by ipanova — APPROVED** [review]
looks right to me!

**Review by dkliban — APPROVED** [review]

**Pull request merged**